File: backend/models/job.py
"""Job model for background job tracking.

This model provides persistent storage for background jobs in PostgreSQL,
enabling job tracking, filtering, and historical job data retention.

Jobs go through states: QUEUED -> RUNNING -> COMPLETED/FAILED/CANCELLED
"""

from __future__ import annotations

from datetime import UTC, datetime
from enum import Enum
from typing import Any

from sqlalchemy import CheckConstraint, DateTime, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from backend.models.camera import Base


class JobStatus(str, Enum):
    """Status of a background job.

    Represents the lifecycle states of a job:
    - QUEUED: Job is waiting to be processed
    - RUNNING: Job is currently executing
    - COMPLETED: Job finished successfully
    - FAILED: Job encountered an error
    - CANCELLED: Job was cancelled by user request
    """

    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"

    def __str__(self) -> str:
        """Return string representation of job status."""
        return self.value


class Job(Base):
    """Job model representing a background job.

    Jobs are created for long-running background operations such as exports,
    imports, cleanup tasks, and backups. This model provides persistent
    storage for job tracking with progress updates, results, and error handling.

    Attributes:
        id: Unique job identifier (UUID string)
        job_type: Type of job (e.g., 'export', 'cleanup', 'backup', 'import')
        status: Current job status (queued, running, completed, failed, cancelled)
        queue_name: Name of the queue this job is assigned to
        priority: Job priority (0 = highest, 4 = lowest)
        created_at: Timestamp when job was created
        started_at: Timestamp when job started running
        completed_at: Timestamp when job finished (success or failure)
        progress_percent: Progress percentage (0-100)
        current_step: Description of current processing step
        result: JSON result data for completed jobs
        error_message: Error message for failed jobs
        error_traceback: Full error traceback for debugging
        attempt_number: Current attempt number (for retries)
        max_attempts: Maximum number of retry attempts
        next_retry_at: Timestamp for next retry attempt
    """

    __tablename__ = "jobs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True)
    job_type: Mapped[str] = mapped_column(String(50), nullable=False)
    status: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default=JobStatus.QUEUED.value,
    )
    queue_name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    priority: Mapped[int] = mapped_column(Integer, nullable=False, default=2)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(UTC),
    )
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Progress tracking
    progress_percent: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    current_step: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # Result and error handling
    result: Mapped[dict[str, Any] | None] = mapped_column(JSONB, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    error_traceback: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Retry handling
    attempt_number: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    max_attempts: Mapped[int] = mapped_column(Integer, nullable=False, default=3)
    next_retry_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    # Indexes for common queries
    __table_args__ = (
        Index("idx_jobs_status", "status"),
        Index("idx_jobs_job_type", "job_type"),
        Index("idx_jobs_created_at", "created_at"),
        Index("idx_jobs_queue_name", "queue_name"),
        Index("idx_jobs_priority", "priority"),
        # Composite index for filtered listing queries
        Index("idx_jobs_status_created_at", "status", "created_at"),
        Index("idx_jobs_job_type_status", "job_type", "status"),
        # BRIN index for time-series queries on created_at (append-only data)
        Index(
            "ix_jobs_created_at_brin",
            "created_at",
            postgresql_using="brin",
        ),
        # CHECK constraints
        CheckConstraint(
            "status IN ('queued', 'running', 'completed', 'failed', 'cancelled')",
            name="ck_jobs_status",
        ),
        CheckConstraint(
            "progress_percent >= 0 AND progress_percent <= 100",
            name="ck_jobs_progress_range",
        ),
        CheckConstraint(
            "priority >= 0 AND priority <= 4",
            name="ck_jobs_priority_range",
        ),
        CheckConstraint(
            "attempt_number >= 1",
            name="ck_jobs_attempt_number_min",
        ),
        CheckConstraint(
            "max_attempts >= 1",
            name="ck_jobs_max_attempts_min",
        ),
        CheckConstraint(
            "completed_at IS NULL OR completed_at >= created_at",
            name="ck_jobs_completed_after_created",
        ),
        CheckConstraint(
            "started_at IS NULL OR started_at >= created_at",
            name="ck_jobs_started_after_created",
        ),
    )

    @property
    def is_active(self) -> bool:
        """Check if the job is still active (queued or running).

        Returns:
            True if the job is queued or running, False otherwise.
        """
        return self.status in (JobStatus.QUEUED.value, JobStatus.RUNNING.value)

    @property
    def is_finished(self) -> bool:
        """Check if the job has finished (completed, failed, or cancelled).

        Returns:
            True if the job has finished, False otherwise.
        """
        return self.status in (
            JobStatus.COMPLETED.value,
            JobStatus.FAILED.value,
            JobStatus.CANCELLED.value,
        )

    @property
    def can_retry(self) -> bool:
        """Check if the job can be retried.

        A job can be retried if it has failed and has remaining attempts.

        Returns:
            True if the job can be retried, False otherwise.
        """
        return self.status == JobStatus.FAILED.value and self.attempt_number < self.max_attempts

    @property
    def duration_seconds(self) -> float | None:
        """Calculate the duration of the job in seconds.

        Returns:
            Duration in seconds if both started_at and completed_at are set,
            None otherwise.
        """
        if self.started_at is None or self.completed_at is None:
            return None
        return (self.completed_at - self.started_at).total_seconds()

    def start(self) -> None:
        """Mark the job as started/running.

        Sets the status to RUNNING and records the start timestamp.
        """
        self.status = JobStatus.RUNNING.value
        self.started_at = datetime.now(UTC)

    def complete(self, result: dict[str, Any] | None = None) -> None:
        """Mark the job as completed.

        Sets the status to COMPLETED, progress to 100%, and records
        the completion timestamp and optional result.

        Args:
            result: Optional result data to store with the job.
        """
        self.status = JobStatus.COMPLETED.value
        self.progress_percent = 100
        self.completed_at = datetime.now(UTC)
        self.current_step = "Completed"
        if result is not None:
            self.result = result

    def fail(self, error_message: str, error_traceback: str | None = None) -> None:
        """Mark the job as failed.

        Sets the status to FAILED and records the error information.

        Args:
            error_message: Human-readable error message.
            error_traceback: Optional full error traceback for debugging.
        """
        self.status = JobStatus.FAILED.value
        self.completed_at = datetime.now(UTC)
        self.error_message = error_message
        self.error_traceback = error_traceback
        self.current_step = (
            f"Failed: {error_message[:50]}..."
            if len(error_message) > 50
            else f"Failed: {error_message}"
        )

    def cancel(self) -> None:
        """Mark the job as cancelled.

        Can only cancel jobs that are queued or running.
        """
        if self.is_active:
            self.status = JobStatus.CANCELLED.value
            self.completed_at = datetime.now(UTC)
            self.current_step = "Cancelled by user"
            self.error_message = "Job cancelled by user request"

    def update_progress(self, progress_percent: int, current_step: str | None = None) -> None:
        """Update the job progress.

        Args:
            progress_percent: New progress value (clamped to 0-100).
            current_step: Optional description of current processing step.
        """
        self.progress_percent = max(0, min(100, progress_percent))
        if current_step is not None:
            self.current_step = current_step

    def prepare_retry(self) -> None:
        """Prepare the job for a retry attempt.

        Increments the attempt number and resets the status to QUEUED.
        Should only be called when can_retry is True.
        """
        if self.can_retry:
            self.attempt_number += 1
            self.status = JobStatus.QUEUED.value
            self.started_at = None
            self.completed_at = None
            self.error_message = None
            self.error_traceback = None
            self.progress_percent = 0
            self.current_step = None

    def __repr__(self) -> str:
        return (
            f"<Job(id={self.id!r}, job_type={self.job_type!r}, "
            f"status={self.status!r}, progress={self.progress_percent}%)>"
        )

File: frontend/src/types/job.ts
/**
 * Job Types - Extended types for job tracking and display
 *
 * These types supplement the generated API types with frontend-specific
 * type definitions and utilities for job management.
 *
 * @module types/job
 * @see NEM-3593
 */

import type {
  JobResponse,
  JobDetailResponse,
  JobStatusEnum,
  components,
} from './generated';

// Re-export generated types for convenience
export type { JobResponse, JobDetailResponse, JobStatusEnum };

// Extract nested types from components
export type JobRetryInfo = components['schemas']['JobRetryInfo'];
export type JobProgressDetail = components['schemas']['JobProgressDetail'];
export type JobTiming = components['schemas']['JobTiming'];
export type JobMetadata = components['schemas']['JobMetadata'];

/**
 * Full Job type combining list and detail response fields.
 * This is a union type that works for both list items and detail views.
 *
 * Use this when you need a flexible type that can handle both
 * JobResponse (from list endpoints) and JobDetailResponse (from detail endpoint).
 */
export type Job = JobResponse | JobDetailResponse;

/**
 * Job type as displayed in the UI, with normalized field names.
 *
 * This interface provides a consistent structure for job display,
 * mapping from both JobResponse and JobDetailResponse fields.
 */
export interface JobDisplayData {
  /** Unique job identifier */
  id: string;
  /** Type of job (e.g., 'export', 'cleanup', 'backup', 'import') */
  job_type: string;
  /** Current job status */
  status: JobStatusEnum;
  /** Name of the queue this job is assigned to (if available) */
  queue_name?: string | null;
  /** Job priority (0 = highest, 4 = lowest) */
  priority?: number;
  /** ISO timestamp when job was created */
  created_at: string;
  /** ISO timestamp when job started running */
  started_at?: string | null;
  /** ISO timestamp when job finished */
  completed_at?: string | null;
  /** Progress percentage (0-100) */
  progress_percent: number;
  /** Description of current processing step */
  current_step?: string | null;
  /** Job result data (if completed) */
  result?: unknown;
  /** Error message (if failed) */
  error_message?: string | null;
  /** Full error traceback for debugging */
  error_traceback?: string | null;
  /** Current attempt number (for retries) */
  attempt_number?: number;
  /** Maximum number of retry attempts */
  max_attempts?: number;
  /** ISO timestamp for next retry attempt */
  next_retry_at?: string | null;
  /** Human-readable status message */
  message?: string | null;
}

/**
 * Convert a JobResponse to JobDisplayData.
 * JobResponse has simpler fields from the list endpoint.
 */
export function jobResponseToDisplayData(job: JobResponse): JobDisplayData {
  return {
    id: job.job_id,
    job_type: job.job_type,
    status: job.status,
    created_at: job.created_at,
    started_at: job.started_at,
    completed_at: job.completed_at,
    progress_percent: job.progress,
    error_message: job.error,
    result: job.result,
    message: job.message,
  };
}

/**
 * Convert a JobDetailResponse to JobDisplayData.
 * JobDetailResponse has more detailed fields from the detail endpoint.
 */
export function jobDetailToDisplayData(job: JobDetailResponse): JobDisplayData {
  return {
    id: job.id,
    job_type: job.job_type,
    status: job.status,
    queue_name: job.queue_name,
    priority: job.priority,
    created_at: job.timing.created_at,
    started_at: job.timing.started_at,
    completed_at: job.timing.completed_at,
    progress_percent: job.progress.percent,
    current_step: job.progress.current_step,
    error_message: job.error,
    result: job.result,
    attempt_number: job.retry_info.attempt_number,
    max_attempts: job.retry_info.max_attempts,
    next_retry_at: job.retry_info.next_retry_at,
  };
}

/**
 * Type guard to check if a job is a JobDetailResponse.
 */
export function isJobDetailResponse(job: Job): job is JobDetailResponse {
  return 'timing' in job && 'retry_info' in job && 'progress' in job && typeof job.progress === 'object';
}

/**
 * Type guard to check if a job is a JobResponse.
 */
export function isJobResponse(job: Job): job is JobResponse {
  return 'job_id' in job && 'progress' in job && typeof job.progress === 'number';
}

/**
 * Convert any Job type to JobDisplayData.
 */
export function toJobDisplayData(job: Job): JobDisplayData {
  if (isJobDetailResponse(job)) {
    return jobDetailToDisplayData(job);
  }
  return jobResponseToDisplayData(job);
}

/**
 * Check if a job has retry information.
 */
export function hasRetryInfo(job: Job): boolean {
  return isJobDetailResponse(job) && job.retry_info.max_attempts > 1;
}

/**
 * Check if a job can be retried (failed and has remaining attempts).
 */
export function canRetry(job: Job): boolean {
  if (!isJobDetailResponse(job)) return false;
  return (
    job.status === 'failed' && job.retry_info.attempt_number < job.retry_info.max_attempts
  );
}

/**
 * Check if a job has a detailed error traceback.
 * Note: Traceback is only available in JobDetailResponse through the error field.
 */
export function hasErrorTraceback(job: Job): boolean {
  return !!job.error && job.error.includes('\n');
}

/**
 * Format retry info for display.
 */
export function formatRetryInfo(retryInfo: JobRetryInfo): string {
  return `Attempt ${retryInfo.attempt_number} of ${retryInfo.max_attempts}`;
}

/**
 * Get the short error message (first line only).
 */
export function getShortErrorMessage(error: string | null | undefined): string | null {
  if (!error) return null;
  const firstLine = error.split('\n')[0];
  return firstLine.length > 100 ? `${firstLine.substring(0, 100)}...` : firstLine;
}

/**
 * Priority level labels for display.
 */
export const PRIORITY_LABELS: Record<number, string> = {
  0: 'Highest',
  1: 'High',
  2: 'Normal',
  3: 'Low',
  4: 'Lowest',
};

/**
 * Get priority label for a priority value.
 */
export function getPriorityLabel(priority: number): string {
  return PRIORITY_LABELS[priority] ?? `Priority ${priority}`;
}
